src/main/java/com/facebook/openwifirrm/ucentral/models/AggregatedState.java

```diff
@@ -0,0 +1,241 @@
+	public int ackSignal;
+	public int ackSignalAvg;
+	public Radio radio;
+
+	/** Constructor with no args */
+	public AggregatedState() {
+		this.rxRate = new AggregatedRate();
+		this.txRate = new AggregatedRate();
+		this.rssi = new ArrayList<>();
+		this.radio = new Radio();
+	}
+
+	/** Construct from Aggregatedstate */
+	public AggregatedState(AggregatedState state) {
+		this.rxRate = new AggregatedRate();
+		this.txRate = new AggregatedRate();
+		this.rssi = new ArrayList<>();
+		add(state);
+	}
+
+	/** Construct from Association and radio */
+	public AggregatedState(
+		Association association,
+		Map<String, Integer> radioInfo
+	) {
+		this.rxRate = new AggregatedRate();
+		this.txRate = new AggregatedRate();
+		this.rssi = new ArrayList<>();
+
+		this.bssid = association.bssid;
+		this.station = association.station;
+		this.connected = association.connected;
+		this.inactive = association.inactive;
+		this.rssi.add(association.rssi);
+		this.rxBytes = association.rx_bytes;
+		this.rxPackets = association.rx_packets;
+		this.rxRate.add(association.rx_rate);
+		this.txBytes = association.tx_bytes;
+		this.txDuration = association.tx_duration;
+		this.txFailed = association.tx_failed;
+		this.txPackets = association.tx_packets;
+		this.txRate.add(association.tx_rate);
+		this.txRetries = association.tx_retries;
+		this.ackSignal = association.ack_signal;
+		this.ackSignalAvg = association.ack_signal_avg;
+		this.radio = new Radio(radioInfo);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(bssid, station, radio);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj) {
+			return true;
+		}
+		if (obj == null) {
+			return false;
+		}
+		if (getClass() != obj.getClass()) {
+			return false;
+		}
+
+		AggregatedState other = (AggregatedState) obj;
+
+		return bssid == other.bssid && station == other.station &&
+			connected == other.connected && inactive == other.inactive && rssi
+				.equals(other.rssi) &&
+			rxBytes == other.rxBytes && rxBytes == other.rxPackets &&
+			rxRate.bitRate == other.rxRate.bitRate &&
+			rxRate.chWidth == other.rxRate.chWidth && rxRate.mcs
+				.equals(other.rxRate.mcs) &&
+			txBytes == other.txBytes && txDuration == other.txDuration &&
+			txFailed == other.txFailed && txPackets == other.txPackets &&
+			txRate.bitRate == other.txRate.bitRate &&
+			txRate.chWidth == other.txRate.chWidth && txRate.mcs
+				.equals(other.txRate.mcs) &&
+			txRetries == other.txRetries && ackSignal == other.ackSignal &&
+			ackSignalAvg == other.ackSignalAvg && radio.equals(other.radio);
+	}
+
+	/**
+	 * Add an AggregatedState to this AggregatedState. Succeed only when the two
+	 * matches in hashCode.
+	 *
+	 * @param state input AggregatedState
+	 * @return boolean true if the two matches in bssid, station, channel,
+	 *         channel_width and tx_power
+	 */
+	public boolean add(AggregatedState state) {
+		if (
+			(bssid == null && station == null && radio == null) ||
```

Comment:
I don't see a `null` ever passed in, except potentially through the constructor from `addStateToAggregation`. If that could be null there, I think we should do a null-check there (like `if (association.bssid == null) {continue;}`) and then remove the null check here. Same with `station` and `radio`. Then we can have a guarantee that `bssid`, `station`, and `radio` are never null (and that would be good to put in their javadoc)

Reply:
Construct from AggregatedState, in this case `(bssid == null && station == null && radio == null)` could be true.